connectors/sources/s3.py

```diff
@@ -35,57 +37,128 @@
     AWS_ENDPOINT = None
 
 
-class S3DataSource(BaseDataSource):
-    """Amazon S3"""
-
-    name = "Amazon S3"
-    service_type = "s3"
+class S3Client:
+    """Amazon S3 client to handle method calls made to S3"""
 
     def __init__(self, configuration):
-        """Set up the connection to the Amazon S3.
-
-        Args:
-            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
-        """
-        super().__init__(configuration=configuration)
+        self.configuration = configuration
         self.session = aioboto3.Session()
         set_extra_logger(aws_logger, log_level=logging.DEBUG, prefix="S3")
         set_extra_logger("aioboto3.resources", log_level=logging.INFO, prefix="S3")
-        self.bucket_list = []
-        self.buckets = self.configuration["buckets"]
         self.config = AioConfig(
             read_timeout=self.configuration["read_timeout"],
             connect_timeout=self.configuration["connect_timeout"],
             retries={"max_attempts": self.configuration["max_attempts"]},
         )
+        self.clients = {}
+        self.client_context = []
+
+    async def client(self, region=None):
+        """This method creates context manager and client session object for s3.
+        Args:
+            region (str): Name of bucket region. Defaults to None
+        """
+        region_name = region if region else "default"
+
+        if region_name in self.clients:
+            return self.clients[region_name]
 
-    @asynccontextmanager
-    async def client(self, **kwargs):
-        """This method creates client object."""
         if AWS_ENDPOINT is not None:
             logger.debug(f"Creating a session against {AWS_ENDPOINT}")
 
-        async with self.session.client(
-            service_name="s3", config=self.config, endpoint_url=AWS_ENDPOINT, **kwargs
+        # AsyncExitStack, supports asynchronous context managers, used to create client using enter_async_context and
+        # these context manager will be stored in client_context list also client will be stored in clients dict with their region
+        s3_context_stack = AsyncExitStack()
+        s3_client = await s3_context_stack.enter_async_context(
+            self.session.client(
+                service_name="s3",
+                config=self.config,
+                endpoint_url=AWS_ENDPOINT,
+                region_name=region,
+            )
+        )
+        self.client_context.append(s3_context_stack)
+
+        self.clients[region_name] = s3_client
+        return self.clients[region_name]
+
+    async def close_client(self):
+        """Closes unclosed client session"""
+        for context in self.client_context:
+            await context.aclose()
+
+    async def fetch_buckets(self):
```

Comment:
should this take a `region` param, so that we can list buckets from any region, and not just from default? Or is that currently working to list all buckets from all regions?

Reply:
@seanstory for fetching the list of all buckets, it is not required to pass region.
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/list_buckets.html